include/aws/crt/Variant.h

```diff
@@ -0,0 +1,559 @@
+                return *value;
+            }
+
+            template <typename T, EnableIfOtherIsThisVariantAlternative<T> = 1> const T *get_if() const
+            {
+                if (holds_alternative<T>())
+                {
+                    T *value = reinterpret_cast<T *>(m_storage);
+                    return value;
+                }
+                else
+                {
+                    return nullptr;
+                }
+            }
+
+            template <std::size_t Index> auto get() const -> const typename ThisVariantAlternative<Index>::type &
+            {
+                static_assert(Index < AlternativeCount, "Unknown alternative index to get");
+                AWS_ASSERT(Index == m_index);
+                using AlternativeT = typename ThisVariantAlternative<Index>::type;
+                const AlternativeT *ret = reinterpret_cast<const AlternativeT *>(m_storage);
+                return *ret;
+            }
+
+            /* This is just a templated way to say
+             * "int*" for
+             * a VariantAlternative<0, Variant<int, char, long>()>*/
+            template <std::size_t Index>
+            using RawAlternativePointerT = typename std::add_pointer<ThisVariantAlternative<Index>>::type;
+
+            template <std::size_t Index> RawAlternativePointerT<Index> *get_if()
+            {
+                static_assert(Index < AlternativeCount, "Unknown alternative index to get");
+                if (holds_alternative<Index>())
+                {
+                    using AlternativePtrT = RawAlternativePointerT<Index>;
+                    AlternativePtrT value = reinterpret_cast<AlternativePtrT>(m_storage);
+                    return value;
+                }
+                else
+                {
+                    return nullptr;
+                }
+            }
+
+            template <std::size_t Index> const RawAlternativePointerT<Index> *get_if() const
+            {
+                static_assert(Index < AlternativeCount, "Unknown alternative index to get");
+                if (holds_alternative<Index>())
+                {
+                    using AlternativePtrT = RawAlternativePointerT<Index>;
+                    AlternativePtrT value = reinterpret_cast<AlternativePtrT>(m_storage);
+                    return value;
+                }
+                else
+                {
+                    return nullptr;
+                }
+            }
+
+            std::size_t index() const { return m_index; }
+
+            ~Variant() { Destroy(); }
+
+          private:
+            static constexpr std::size_t STORAGE_SIZE = VariantDetail::ParameterPackSize::GetMaxSizeOf<Ts...>();
+
+            alignas(VariantDetail::ParameterPackSize::AlignAsPack<Ts...>()) char m_storage[STORAGE_SIZE];
+            VariantDetail::Index::VariantIndex m_index = -1;
+
+            void Destroy()
+            {
+                if (m_index != -1)
```

Comment:
I tried to eliminate all cases where user can set -1 type in the variant.
There are still corner cases (explicit d-tor call) or UB where this may happen - I've added asserts or fatal asserts accordingly.